### PR TITLE
`ValidNext`

### PR DESCRIPTION
**Pull request by mistermoe**
### Summary
* Added [`tbdex.ParseMessage`](https://github.com/TBD54566975/tbdex-go/blob/validnext/tbdex/tbdex.go#L77-L124) which can be used to parse a message when the kind isn't known upfront
* Added [`tbdex.DecodeMessage`](https://github.com/TBD54566975/tbdex-go/blob/validnext/tbdex/tbdex.go#L29-L75) which can be used to decode a message when the kind isn't known upfront
* Added `ValidNext()` to each message kind in order to help with exchange validation

> [!NOTE]
>  Most of this PR is moving files around to avoid circular dependencies.

---

### Motivation
There are many scenarios on both client and server side wherein a message is received and the kind is not known up front. Specifically, when requesting an entire exchange on the client side, or receiving a new message to add to an exchange. In these scenarios, it's also necessary to validate whether a given message kind is valid relative to the current state of a given exchange

---

### Usage

```go
msg, err := tbdex.ParseMessage(runtimeBytes)
order, ok := msg.(order.Order)
if !ok {
    panic("nope")
}
```

> [!NOTE]
> `ParseMessage` and `DecodeMessage` return the [`Message`](https://github.com/TBD54566975/tbdex-go/blob/validnext/tbdex/tbdex.go#L14-L23) interface. Returning interfaces can be seen as an anti-pattern in certain scenarios where doing so introduces unnecessary complexity or unnecessarily reduces type fidelity when other approaches can achieve the same intended result without losing strong safety. 
>
> It is however in some scenarios perfectly acceptable particularly when doing so is motivated by a real need vs. preemptively.  "Don’t design with interfaces, discover them." is referenced quite often within the go community. Returning interfaces exists within go itself e.g. `error`, `io.Reader`, `io.Writer` etc.
>
> for our use case specifically it's especially helpful in scenarios where a message needs to be decoded or parsed but the kind is not known upfront. There's really no way for go to magically know what strong type to use at runtime automatically.

### CODE REVIEW

File: tbdex/validator/validator_test.go [hunk 1]
```diff
@@ -0,0 +1,6 @@
+package validator_test
+
+// func TestValidate_Invalid(t *testing.T) {
+// 	err := tbdex.Validate(tbdex.TypeResource, []byte(`{"foo": "bar"}`))
+// 	assert.Error(t, err)
+// }
```
Comment (diehuxx) [inline tbdex/validator/validator_test.go]: is this TODO in this PR or a follow up?

**Review by diehuxx — APPROVED** [review]

File: tbdex/tbdex.go [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+type msg struct {
+	Metadata message.Metadata
+}
+
+// DecodeMessage unmarshals a message. It uses the metadata kind to determine the type of message.
+//
+// # Note
+//
+// unmarshaling includes validation
+func DecodeMessage(input []byte) (Message, error) {
```
Comment (phoebe-lew) [inline tbdex/tbdex.go]: Nit: Why call `decode` instead of `unmarshal`?

File: tbdex/tbdex.go [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+	var m msg
+	err := json.Unmarshal(input, &m)
+
+	if err != nil {
+		return nil, fmt.Errorf("failed to unmarshal partial message to determine kind: %w", err)
+	}
```
Comment (phoebe-lew) [inline tbdex/tbdex.go]:
```suggestion
	if err := json.Unmarshal(input, &m); err != nil {
		return nil, fmt.Errorf("failed to unmarshal partial message to determine kind: %w", err)
	}
```

File: tbdex/validator/validator_test.go [hunk 4]
```diff
@@ -0,0 +1,6 @@
+package validator_test
+
+// func TestValidate_Invalid(t *testing.T) {
```
Comment (phoebe-lew) [inline tbdex/validator/validator_test.go]: delete?

File: tbdex/balance/balance.go [hunk 5]
```diff
@@ -15,9 +17,9 @@ const Kind = "balance"
 
 // Balance is a resource to communicate the amounts of each currency held by the PFI on behalf of its customer.
 type Balance struct {
-	Metadata  tbdex.ResourceMetadata `json:"metadata,omitempty"`
-	Data      Data                   `json:"data,omitempty"`
-	Signature string                 `json:"signature,omitempty"`
+	Metadata  resource.Metadata `json:"metadata,omitempty"`
```
Comment (phoebe-lew) [inline tbdex/balance/balance.go:20]: I think this reads nicer than the old devex fwiw!

**Review by phoebe-lew — APPROVED** [review]

File: tbdex/tbdex.go (hunk 2 above)
Comment (mistermoe) [reply tbdex/tbdex.go]: Good point. Can do will do! Was sad that I couldn't use `UnmarshalJSON` on an interface but made sense as to why

File: tbdex/validator/validator_test.go (hunk 4 above)
Comment (mistermoe) [reply tbdex/validator/validator_test.go]: Yee

File: tbdex/tbdex.go (hunk 2 above)
Comment (mistermoe) [reply tbdex/tbdex.go]: done! https://github.com/TBD54566975/tbdex-go/commit/f791e147fd90daa88eb5896f8646e27571bf817b

File: tbdex/validator/validator_test.go (hunk 4 above)
Comment (mistermoe) [reply tbdex/validator/validator_test.go]: oops. i dont even remember commenting this out. uncommented: https://github.com/TBD54566975/tbdex-go/commit/8622cccbcbf1cac735c111df1dea32e5483bdd58

File: tbdex/validator/validator_test.go (hunk 1 above)
Comment (mistermoe) [reply tbdex/validator/validator_test.go]: https://github.com/TBD54566975/tbdex-go/commit/8622cccbcbf1cac735c111df1dea32e5483bdd58

**Pull request merged**